datafusion/src/row/writer.rs

```diff
@@ -0,0 +1,322 @@
+            }
+        }
+    };
+}
+
+/// Reusable row writer backed by Vec<u8>
+pub struct RowWriter {
+    data: Vec<u8>,
+    field_count: usize,
+    row_width: usize,
+    null_width: usize,
+    values_width: usize,
+    varlena_width: usize,
+    varlena_offset: usize,
+    field_offsets: Vec<usize>,
+}
+
+impl RowWriter {
+    /// new
+    pub fn new(schema: &Arc<Schema>) -> Self {
+        assert!(supported(schema));
+        let field_count = schema.fields().len();
+        let null_width = bytes_for(field_count);
+        let (field_offsets, values_width) = get_offsets(null_width, schema);
+        let mut init_capacity = estimate_row_width(null_width, schema);
+        if !fixed_size(schema) {
+            // double the capacity to avoid repeated resize
+            init_capacity *= 2;
+        }
+        Self {
+            data: vec![0; init_capacity],
+            field_count,
+            row_width: 0,
+            null_width,
+            values_width,
+            varlena_width: 0,
+            varlena_offset: null_width + values_width,
+            field_offsets,
+        }
+    }
+
+    /// Reset the row writer state for new tuple
+    pub fn reset(&mut self) {
+        self.data.fill(0);
+        self.row_width = 0;
+        self.varlena_width = 0;
+        self.varlena_offset = self.null_width + self.values_width;
+    }
+
+    #[inline]
+    fn assert_index_valid(&self, idx: usize) {
+        assert!(idx < self.field_count);
+    }
+
+    fn set_null_at(&mut self, idx: usize) {
+        let null_bits = &mut self.data[0..self.null_width];
+        set_bit(null_bits, idx, false)
+    }
+
+    fn set_non_null_at(&mut self, idx: usize) {
+        let null_bits = &mut self.data[0..self.null_width];
+        set_bit(null_bits, idx, true)
+    }
+
+    fn set_bool(&mut self, idx: usize, value: bool) {
+        self.assert_index_valid(idx);
+        let offset = self.field_offsets[idx];
+        self.data[offset] = if value { 1 } else { 0 };
+    }
+
+    fn set_u8(&mut self, idx: usize, value: u8) {
+        self.assert_index_valid(idx);
+        let offset = self.field_offsets[idx];
+        self.data[offset] = value;
+    }
+
+    fn_set_idx!(u16, 2);
+    fn_set_idx!(u32, 4);
+    fn_set_idx!(u64, 8);
+    fn_set_idx!(i16, 2);
+    fn_set_idx!(i32, 4);
+    fn_set_idx!(i64, 8);
+    fn_set_idx!(f32, 4);
+    fn_set_idx!(f64, 8);
+
+    fn set_i8(&mut self, idx: usize, value: i8) {
+        self.assert_index_valid(idx);
+        let offset = self.field_offsets[idx];
+        self.data[offset] = value.to_le_bytes()[0];
+    }
+
+    fn set_date32(&mut self, idx: usize, value: i32) {
+        set_idx!(4, self, idx, value)
+    }
+
+    fn set_date64(&mut self, idx: usize, value: i64) {
+        set_idx!(8, self, idx, value)
+    }
+
+    fn set_offset_size(&mut self, idx: usize, size: usize) {
+        let offset_and_size: u64 = (self.varlena_offset << 32 | size) as u64;
```

Comment:
This should probably panic if `size` or `verlana_offset` are too large, on a related note - perhaps they should be `u64`?